src/librustc_mir/transform/add_retag.rs

```diff
@@ -0,0 +1,169 @@
+
+    match *place {
+        Local { .. } => true,
+        Promoted(_) |
+        Static(_) => false,
+        Projection(ref proj) => {
+            match proj.elem {
+                ProjectionElem::Deref |
+                ProjectionElem::Index(_) =>
+                    // Which place these point to depends on external circumstances
+                    // (a local storing the array index, the current value of
+                    // the projection base), so we stop tracking here.
+                    false,
+                _ => is_local(&proj.base),
+            }
+        }
+    }
+}
+
+/// Determine whether this type has a reference in it, recursing below compound types but
+/// not below references.
+fn has_reference<'a, 'gcx, 'tcx>(ty: Ty<'tcx>, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
+    match ty.sty {
+        // Primitive types that are not references
+        ty::Bool | ty::Char |
+        ty::Float(_) | ty::Int(_) | ty::Uint(_) |
+        ty::RawPtr(..) | ty::FnPtr(..) |
+        ty::Str | ty::FnDef(..) | ty::Never =>
+            false,
+        // References
+        ty::Ref(..) => true,
+        ty::Adt(..) if ty.is_box() => true,
+        // Compound types
+        ty::Array(ty, ..) | ty::Slice(ty) =>
+            has_reference(ty, tcx),
+        ty::Tuple(tys) =>
+            tys.iter().any(|ty| has_reference(ty, tcx)),
+        ty::Adt(adt, substs) =>
+            adt.variants.iter().any(|v| v.fields.iter().any(|f|
+                has_reference(f.ty(tcx, substs), tcx)
+            )),
+        // Conservative fallback
+        _ => true,
+    }
+}
+
+impl MirPass for AddRetag {
+    fn run_pass<'a, 'tcx>(&self,
+                          tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                          _src: MirSource,
+                          mir: &mut Mir<'tcx>)
+    {
+        if !tcx.sess.opts.debugging_opts.mir_emit_retag {
+            return;
+        }
+        let (span, arg_count) = (mir.span, mir.arg_count);
+        let (basic_blocks, local_decls) = mir.basic_blocks_and_local_decls_mut();
+        let needs_retag = |place: &Place<'tcx>| {
+            is_local(place) && has_reference(place.ty(&*local_decls, tcx).to_ty(tcx), tcx)
+        };
+
+        // PART 1
+        // Retag arguments at the beginning of the start block.
+        {
+            let source_info = SourceInfo {
+                scope: OUTERMOST_SOURCE_SCOPE,
+                span: span, // FIXME: Consider using just the span covering the function
+                            // argument declaration.
+            };
+            // Gather all arguments, skip return value.
+            let places = local_decls.iter_enumerated().skip(1).take(arg_count)
+                    .map(|(local, _)| Place::Local(local))
+                    .filter(needs_retag)
+                    .collect::<Vec<_>>();
+            // Emit their retags.
+            basic_blocks[START_BLOCK].statements.splice(0..0,
+                places.into_iter().map(|place| Statement {
```

Comment:
The old system used to have multiple places in one validation statement. What is the reason this got changed?

Reply:
I wanted to avoid the allocation and indirection. The old system had two cases where we had groups of things to validate (function prologue and before a function call); now there is only one so it didn't seem worth it.

I can make it a `Vec` again if you prefer.

Reply:
I was just curious about the reason